Support AMD SEV-SNP on AWS

fangge1212 opened the pull request:
AMD SEV-SNP is one of the confidential computing technologies. This commit adds support for AMD SEV-SNP on AWS, so users can utilize the confidential computing on the cluster nodes.

Upstream CAPA PR: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5605

bgartzi commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -17,6 +17,9 @@ type AWSMachineProviderConfig struct {
+	// cpuOptions is the set of cpu options for the instance, where you can enable/disable AMD SEV-SNP on the instance.
+	// +optional
+	CpuOptions *CpuOptions `json:"cpuOptions,omitempty"`
 	// tags is the set of tags to add to apply to an instance, in addition to the ones
```
Hi @fangge1212, would it make sense to use a field named `ConfidentialComputing` (or something similar) that could be `Disabled` or `AmdSevSnp` instead?

In case AWS supports other confidential computing technologies in the future, that would be easier to extend than this approach IMO. As it would require adding another new field to the API instead of a new value to the enum.

fangge1212 replied:
Same as https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5598, waiting for more opinions before chaning it

fangge1212 commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -109,6 +112,31 @@ type AWSMachineProviderConfig struct {
+)
+
+// CpuOptions defines the cpu options for the instance.
+type CpuOptions struct {
```
- If it is an empty struct, cpuOptions=nil will be passed to the AWS API
- I don't get your question. If there are multiple options to configure in the future, then we can configure multiple options.

fangge1212 replied:
Updated to:
```
cpuOptions:
  confidentialComputing: AMDSevSnp
```

fangge1212 commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -17,6 +17,9 @@ type AWSMachineProviderConfig struct {
 	// instanceType is the type of instance to create. Example: m4.xlarge
 	InstanceType string `json:"instanceType"`
+	// cpuOptions is the set of cpu options for the instance.
+	// +optional
```
If unset, no CPU options are passed to the AWS platform and AWS default values are used.

JoelSpeed replied:
Do we know what the default values are currently on AWS?

fangge1212 replied:
The CpuOptions in AWS consists of three fields: 
- amdSevSnp - default value is Disabled.
- coreCount - default value depends on the instance type.
- threadsPerCore - default value depends on the instance type.
Refer to: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/cpu-options-supported-instances-values.html

In this PR, only amdSevSnp is exposed to users. I'm not entirely sure how best to describe this in the API documentation — should we include a link to the AWS documentation?

fangge1212 commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -109,6 +112,33 @@ type AWSMachineProviderConfig struct {
+type CPUOptions struct {
+	// confidentialCompute specifies whether confidential computing should be enabled for the instance,
+	// and, if so, which confidential computing technology to use.
+	// If set to Disabled, the instance will not use confidential computing.
```
Updated

fangge1212 commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -17,6 +17,9 @@ type AWSMachineProviderConfig struct {
 	AMI AWSResourceReference `json:"ami"`
 	// instanceType is the type of instance to create. Example: m4.xlarge
 	InstanceType string `json:"instanceType"`
+	// cpuOptions is the set of cpu options for the instance.
```
Thanks. I updated the doc, please have a look again.

fangge1212 commented:
/retest-required

fangge1212 commented:
/retest

fangge1212 commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -109,6 +112,33 @@ type AWSMachineProviderConfig struct {
+)
+
+// CPUOptions defines the cpu options for the instance.
+type CPUOptions struct {
```
`+kubebuilder:validation:MinProperties=1` makes many tests failed:

everettraven replied:
Looks like that is in progress here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5624

damdo commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -109,6 +113,35 @@ type AWSMachineProviderConfig struct {
+	AWSConfidentialComputePolicyNone AWSConfidentialComputePolicy = "None"
+	// AWSConfidentialComputePolicySEVSNP enables AMD SEV-SNP as the confidential computing technology for the instance.
+	AWSConfidentialComputePolicySEVSNP AWSConfidentialComputePolicy = "AmdSevSnp"
+)
```
`Enabled` is normally for the default value as you can see here on the GCP one:

```
	// When set to Enabled, the machine will be configured as a confidential computing instance with no preference on the confidential compute policy used. In this mode, the platform chooses a default that is subject to change over time. Currently, the default is to use AMD Secure Encrypted Virtualization.
```

Ideally we should try and keep consistency across providers for the same feature

bgartzi replied:
Hi @damdo, @fangge1212!

The problem I see in the AWS case is that AFAICT, it doesn't support the default GCP case: AMD SEV. It only supports AMD SEV-SNP. In that case, we would end up defining `Enabled` as a valid case for AWS, but would have a different behavior than the one from GCP, configuring AMD SEV-SNP machines instead. In my opinion this would be misleading.

This is another discussion, but at the moment the `Enabled` option for GCP is redundant, as `AMDEncryptedVirtualization` achieves the same result. In my opinion, Deprecating `Enabled` in the GCP provider in the future would help getting rid of that redundancy and also having consistency across providers for the same feature, as you mention.

damdo replied:
If we are thinking about deprecating `Enabled` then it is probably ok to leave it out from this one.
cc. @JoelSpeed

everettraven replied:
We can also use the omission of the parent field to do defaulting behavior as well, so just omitting the `cpuOptions` field altogether could mean "no opinion on CPU options, platform defaults"

everettraven replied:
>The current code doesn't support this option: "I don't care, I just want confidential computing to be enabled, I'll let OpenShift default it to something". Do we need this?

Is that behavior you want your users to be able to do? I wouldn't say this is a requirement on my end, but I do think that we should be clear on the intended user workflows to be able to design this API appropriately.

fangge1212 replied:
I don’t think `Platform` is very useful in practice. When a user installs a cluster, they already choose the instance type and region, which directly constrains the available confidential computing technologies. That means the platform can’t meaningfully pick a “default” technology on the user’s behalf.

everettraven commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -17,6 +17,10 @@ type AWSMachineProviderConfig struct {
 	AMI AWSResourceReference `json:"ami"`
 	// instanceType is the type of instance to create. Example: m4.xlarge
 	InstanceType string `json:"instanceType"`
+	// cpuOptions defines CPU-related settings for the instance, including the confidential computing policy.
+	// If unset, no CPU options will be passed to the AWS platform and AWS default CPU options will be applied.
```
How do I know what the "AWS default CPU options" that will be applied are? Are these literally defaults AWS imposes on requests that don't specify these options, or are these defaulted elsewhere?

fangge1212 replied:
Updated the description again to make it more concise: 
` When omitted, this means no opinion and the AWS platform is left to choose a reasonable default.`

fangge1212 commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -17,6 +17,12 @@ type AWSMachineProviderConfig struct {
+	// cpuOptions defines CPU-related settings for the instance, including the confidential computing policy.
+	// If unset, no cpuOptions will be included in the API request to AWS, and the instance will use the default CPU options
+	// applied by AWS for the selected intance type.
+	// +kubebuilder:validation:MinProperties=1
```
Updated.

fangge1212 commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -109,6 +115,35 @@ type AWSMachineProviderConfig struct {
+	// More details can be checked at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/sev-snp.html
+	// When omitted, this means no opinion and the AWS platform is left to choose a reasonable default,
+	// which is subject to change without notice. The current default is Disabled.
+	// +kubebuilder:validation:Enum=Disabled;AMDEncrytedVirtualizationNestedPaging
```
The placement of this line differs between `AWSNetworkInterfaceType` (within the struct) and `MarketType `(on the type definition). I was unsure which pattern to follow, so I selected one approach arbitrarily.

JoelSpeed commented:
API looks good, but cannot merge until the equivalent upstream API is merged into CAPA, please LMK once that is merged

damdo commented:
New CAPA upstream PR for this is here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5605
And at the moment it is blocked on https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5624

JoelSpeed commented:
/hold until upstream PRs have been merged

fangge1212 commented:
/unhold
[Upstream pr](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5605 ) is merged

everettraven commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -109,6 +113,36 @@ type AWSMachineProviderConfig struct {
+type CPUOptions struct {
+	// confidentialCompute specifies whether confidential computing should be enabled for the instance,
+	// and, if so, which confidential computing technology to use.
+	// Valid values are: Disabled, AMDEncryptedVirtualizationNestedPaging
```
Nit: We generally also mention that an allowed value is `omitted` for optional enums.

In practice, omitting this field is for now an invalid configuration, but in the future if new fields are added omitting this field and specifying another one would be valid.

everettraven commented:
@fangge1212 Could you include a link to the associated PR that implements the validations outlined here in the validating webhook this API runs through?

fangge1212 commented:
> Could you include a link to the associated PR that implements the validations outlined here in the validating webhook this API runs through?

Previously, I added instance type validation for the specified confidential computing technology (currently only AMD SEV-SNP). However, the instance type list is hard-coded and needs to be updated manually, so I removed the validation. As a result, there is now no webhook validation for this API change.

everettraven commented:
>Previously, I added instance type validation for the specified confidential computing technology (currently only AMD SEV-SNP). However, the instance type list is hard-coded and needs to be updated manually, so I removed the validation. As a result, there is now no webhook validation for this API change.

AFAIK none of the validations that exist on the markers in the API here are actually enforced unless done through the webhook. I would expect us to have changes in the webhook to reject invalid configurations.

fangge1212 commented:
> > > Previously, I added instance type validation for the specified confidential computing technology (currently only AMD SEV-SNP). However, the instance type list is hard-coded and needs to be updated manually, so I removed the validation. As a result, there is now no webhook validation for this API change.
> > 
> > 
> > AFAIK none of the validations that exist on the markers in the API here are actually enforced unless done through the webhook. I would expect us to have changes in the webhook to reject invalid configurations.
> 
> Ah, I didn't know this. I will add validation in webhook

When I added `+kubebuilder:validation:MinProperties=1` before, I had this error. It seems the validation on the marker is working?
```
    --- FAIL: TestSSHKeyName/SSH_key_name_is_nil_is_valid (0.01s)
        sshkeyname_test.go:89: ValidateCreate() error = AWSMachine.infrastructure.cluster.x-k8
       "machine-9zsqb" is invalid: spec.cpuOptions: Invalid value: 0: spec.cpuOptions in body should have at least 1 properties, wantErr false
```

everettraven commented:
That looks like it is an upstream test. @JoelSpeed knows the nuances here more than I do, but my understanding is that machine configurations for OpenShift are embedded as raw JSON in some other API (i forget which one) and as such they must be programmatically evaluated

everettraven commented:
@fangge1212 Yes, I'll override the failure and approve this PR once I am confident that the appropriate validations are in place on the webhook side. I left a comment on that draft PR regarding a case that looks to be considered valid that should not be based on the current API constraints.

everettraven commented on machine/v1beta1/types_awsprovider.go:
```diff
@@ -109,6 +116,37 @@ type AWSMachineProviderConfig struct {
+	// Valid values are: Disabled, AMDEncryptedVirtualizationNestedPaging and omitted.
+	// When set to Disabled, confidential computing will be disabled for the instance.
+	// When set to AMDEncryptedVirtualizationNestedPaging, AMD SEV-SNP will be used as the confidential computing technology for the instance.
+	// In this case, ensure the following conditions are met:
+	// 1) The selected instance type supports AMD SEV-SNP.
+	// 2) The selected AWS region supports AMD SEV-SNP.
+	// 3) The selected AMI supports AMD SEV-SNP.
+	// More details can be checked at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/sev-snp.html
```
I forgot to ask this originally, but what happens if the user specifies AMD SEV-SNP on an instance type, region, or AMI that does _not_ support this?

Is there some feedback loop that enables a user to recover from this bad state?

fangge1212 replied:
In cluster-api-provider-aws, errors from the EC2 RunInstances API (such as InvalidParameterCombination for unsupported instance types, AMIs, or regions) are caught and wrapped before being returned. These errors propagate up through the reconciliation loop until they surface in the AWSMachine (or higher-level resource) status conditions.

everettraven replied:
Gotcha. So regardless the installer will always return the failure.

@JoelSpeed this is relatively new territory for me as I haven't reviewed a lot of these machine API changes - is there a preference as to failing at configuration read time over during cluster bootstrap? Do you know if we clean up any resources we may or may not have created if we fail bootstrapping? What is the impact to an end-user if we fail during cluster bootstrapping versus validation of the configuration?

everettraven reviewed (approved):
/lgtm

everettraven commented:
@fangge1212 over to you for getting the `verified` label

fangge1212 commented:
/verified later @yalzhang
